### PR TITLE
Add hostname parameter

### PR DESCRIPTION
**Pull request by anemyte**
This is refactoring of https://github.com/prometheus/blackbox_exporter/pull/816
As per [the design doc](https://docs.google.com/document/d/1VwqXi2TOb5KXaZY6Iio7411x64pJao3GusX8MqYsJ2g), the exporter now accept `hostname` parameter, which sets `Host` header for HTTP probes. It also sets TLS server name (if it is not already defined by module configuration) to avoid TLS handshake problems when `target` is an IP address. Setting `Host` with both module configuration and `hostname` parameter results in 400 Bad Request.

### CODE REVIEW

**Comment by anemyte** [comment]
@roidelapluie please look

**Comment by imneov** [comment]
My suggestion is to use the following method:

- Expose as HTTP query param as well: 
```
http://127.0.0.1:9115/probe?target=214.32.12.4&module=http_2xx&HTTP.hostname=prometheus.io
http://127.0.0.1:9115/probe?target=214.32.12.4&module=http_2xx&HTTP.header.field_a=value_a&HTTP.header.field_b=value_b
```
- Expose DNS domain as HTTP query param.
```
http://127.0.0.1:9115/probe?target=8.8.8.8&module=dns&DNS.query_name=prometheus.io
```



In this way, the consistency of the use mode, avoid parameter conflicts, misoperation, and expandability.

**Comment by anemyte** [comment]
@roidelapluie I've added fixes, hope it looks better now. Can you please look again?

File: main.go [hunk 1]
```diff
@@ -150,6 +159,23 @@ func probeHandler(w http.ResponseWriter, r *http.Request, c *config.Config, logg
 	h.ServeHTTP(w, r)
 }
 
+func setHttpHost(hostname string, module *config.Module) error {
```
Comment (roidelapluie) [inline main.go:162]:
```suggestion
func setHTTPHost(hostname string, module *config.Module) error {
```

**Comment by roidelapluie** [comment]
Thanks! Just one nit.

**Review by roidelapluie — APPROVED** [review]

**Comment by roidelapluie** [comment]
Thanks!

**Comment by anemyte** [comment]
> Thanks!

Thank you for the review :)

**Pull request merged**